icefall/context_graph.py

```diff
@@ -0,0 +1,222 @@
+        self.root = ContextState(token=-1, score=0, total_score=0, is_end=False)
+        self.root.fail = self.root
+
+    def _fill_fail(self):
+        """This function fills the fail arc for each trie node, it can be computed
+        in linear time by performing a breadth-first search starting from the root.
+        See https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm for the
+        details of the algorithm.
+        """
+        queue = []
+        for token, node in self.root.next.items():
+            node.fail = self.root
+            queue.append(node)
+        while queue:
+            current_node = queue.pop(0)
+            current_fail = current_node.fail
+            for token, node in current_node.next.items():
+                fail = current_fail
+                if token in current_fail.next:
+                    fail = current_fail.next[token]
+                node.fail = fail
+                queue.append(node)
+
+    def build_context_graph(self, token_ids: List[List[int]]):
+        """Build the ContextGraph from a list of token list.
+        It first build a trie from the given token lists, then fill the fail arc
+        for each trie node.
+
+        See https://en.wikipedia.org/wiki/Trie for how to build a trie.
+
+        Args:
+          token_ids:
+            The given token lists to build the ContextGraph, it is a list of token list,
+            each token list contains the token ids for a word/phrase. The token id
+            could be an id of a char (modeling with single Chinese char) or an id
+            of a BPE (modeling with BPEs).
+        """
+        for tokens in token_ids:
+            node = self.root
+            for i, token in enumerate(tokens):
+                if token not in node.next:
+                    node.next[token] = ContextState(
+                        token=token,
+                        score=self.context_score,
+                        # The total score is the accumulated score from root to current node,
+                        # it will be used to calculate the score of fail arc later.
+                        total_score=node.total_score + self.context_score,
+                        is_end=i == len(tokens) - 1,
+                    )
+                node = node.next[token]
+            self._fill_fail()
+
+    def forward_one_step(
+        self, state: ContextState, token: int
+    ) -> Tuple[float, ContextState]:
+        """Search the graph with given state and token.
+
+        Args:
+          state:
+            The given state (trie node) to start.
+          token:
+            The given token.
+
+        Returns:
+          Return a tuple of score and next state.
+        """
+        # token matched
+        if token in state.next:
+            node = state.next[token]
+            score = node.score
+            # if the matched node is the end of a word/phrase, we will start
+            # from the root for next token.
+            if node.is_end:
```

Comment:
This seems not right, the search set (phrase list) may contain a string A(say "ab") that is a prefix of a string B(say "abc"). Even the 'b' of "ab" is end(algorithmically, we often call this an "output node" instead of "end", which is misleading), the match should be able to continue along the path. Actually, the correct matching behavior should follow "either a match or a failure arc" strictly. I'll recommand some links as references in the main thread later.

Reply:
Yes, I know that the Aho-Corasick does not match this way,  my current implementation can match shortest phrase if some phrases have common prefixs. I tried the original matching method, but it gets worse performance. I suspect it is because the wrong ``fill_fail``. Will re-run the experiments.

Reply:
It is expected to match both phrases like ("ab", "abc"),  the difficulty is tracking the scores, will first make sure the graph constructed is correct, than let's consider the searching.